addon/model.js

```diff
@@ -94,28 +94,53 @@ function resolveReference(reference, store) {
   }
 }
 
-function resolveValue(key, value, modelName, store, schema, model, valueInArray = false) {
-  const schemaInterface = model._internalModel._modelData.schemaInterface;
-
-  // TODO: remove this in favour of computeAttributeReference returning arrays
-  if (schema.isAttributeArrayReference(key, value, modelName, schemaInterface)) {
+function resolveReferenceOrReferences(
+  key,
+  value,
+  reference,
+  modelName,
+  store,
+  schema,
+  schemaInterface
+) {
+  if (Array.isArray(value) || Array.isArray(reference)) {
     return resolveRecordArray(key, value, modelName, store, schema, schemaInterface);
   }
 
-  // TODO: have computeNestedModel return arrays instead
-  if (Array.isArray(value)) {
-    return resolvePlainArray(key, value, modelName, store, schema, model);
-  }
+  return resolveReference(reference, store);
+}
 
+/**
+ * There are two different type of values we have to worry about:
+ * 1. References
+ * 2. Nested Models
+ *
+ * Here is a mapping of input -> output:
+ * 1. Single reference -> resolved reference
+ * 2. Array of references -> RecordArray of resolved references
+ * 3. Single nested model -> EmbeddedMegaMorphicModel
+ * 4. Array of nested models -> array of EmbeddedMegaMorphicModel
+ */
+function resolveValue(key, value, modelName, store, schema, model, valueInArray = false) {
+  const schemaInterface = model._internalModel._modelData.schemaInterface;
+
+  // First check to see if given value is either a reference or an array of references
   let reference = _computeAttributeReference(key, value, modelName, schemaInterface, schema);
   if (reference !== undefined && reference !== null) {
-    if (Array.isArray(reference)) {
-      return reference.map(singleRef => resolveReference(singleRef, store));
-    } else {
-      return resolveReference(reference, store);
-    }
+    return resolveReferenceOrReferences(
+      key,
+      value,
+      reference,
+      modelName,
+      store,
+      schema,
+      schemaInterface
+    );
   }
 
+  if (Array.isArray(value)) {
+    return resolvePlainArray(key, value, modelName, store, schema, model);
+  }
```

Comment:
can we drop this `isArray(value)` check and instead have `schema.computeNestedModel` return an array (ie let userland determine whether this is an array of nested models or not?

If that's at all a pain i'm 👍  on doing that in a separate pr as well

Reply:
I'll definitely work on this.

I'll fix this with a follow up PR